R/tf_estimator.R

```diff
@@ -141,6 +154,7 @@ predict.tf_estimator <- function(object,
                                  predict_keys = NULL,
                                  hooks = NULL,
                                  as_iterable = FALSE,
+                                 simplify = TRUE,
```

Comment:
I think that TRUE should really mean "auto", which is to use a simplifier for canned estimators and to use a simplifier for custom estimators *when it is available". There should be a way for custom estimators to include a simplifier in their definition (either something returned from model_fn or an S3 method, whichever makes the most sense).

Reply:
Could you elaborate a little more why it's necessary to include this in their definition? This simplifier is currently only used in prediction phase. Do you mean we should allow users to specify this in two possible places? It might be better to make it consistent for both canned and custom estimators to specify simplifier in one place.

Reply:
I think there's effectively three behaviors users might want at this level:

`TRUE`: Use the simplifer function associated with an estimator (if any).
`FALSE`: Don't simplify anything; return the 'base' output as from TensorFlow.
A function: Use this function in lieu of any registered simplifier on the estimator. (This use case is somewhat more esoteric)

Reply:
I guess I was just thinking that a "custom" estimator might actually be a library that someone has created for others to use. In that case I don't think the end users should be burdened with having to learn a special syntax (pass a specific simplifier function) to use that estimator the same way they use canned estimators (which get simplified output automatically).

I guess I'm just asking that custom estimators be able to appear to the end user like canned estimators. I realize that if the developer of a custom estimator is the same one calling predict it's just as straightforward to pass a simplify function, but for their end users I think it's better if it's somehow bound to the definition of the custom estimator.

Reply:
I think in most cases the developer of a custom estimator should be the same one calling predict. Otherwise, it would be a canned estimator. Assume there's a `xxx_estimator`, developer should then implement a `predict.xxx_estimator()` with a custom simplifier targeting this new estimator. Then users would only need to call `predict(xxx_estimator_object, simplify = TRUE)`. 

This simplifier is too specific to the prediction phase that it's better not to add it as part of the estimator definition. Estimator definition should only contain the model definitions and parameters. We can think of this simplifier as `"output_fn"` that should be tied to the specific actions on the model (e.g. train, predict, evaluate). 

There could potentially be a `evaluation_simplifier` in the future that we only want to add as part of `evaluate()`. Otherwise in the model definition there will be multiple simplifiers. We can see these as part of different specs (e.g. predict_spec, eval_spec) similar to what's in my `train_and_evaluate()` PR. This will prevent the parameters in the model definition from exploding.

Reply:
Okay, I buy all that. Thanks for bearing with me on this one, I'm now convinced we've got the right approach.